crates/proto/src/rr/dnssec/proof.rs

```diff
@@ -0,0 +1,421 @@
+
+    /// There was a protocol error when looking up DNSSEC records
+    #[error("communication failure for query: {query}: {proto}")]
+    Proto {
+        /// Query that failed
+        query: Query,
+        /// Resons fo the failure
+        proto: ProtoError,
+    },
+
+    /// The RRSIGs for the rrset are not present.
+    ///    It's indeterminate if DS records can't be found
+    ///    It's bogus if the DS records are present
+    #[error("rrsigs are not present for: {name} record_type: {record_type}")]
+    RrsigsNotPresent {
+        /// Name that RRSIGS are missing for
+        name: Name,
+        /// The record type in question
+        record_type: RecordType,
+    },
+
+    /// The RRSIGs could not be verified or failed validation
+    #[error("rrsigs were not able to be verified: {name}, type: {record_type}")]
+    RrsigsUnverified {
+        /// Name that RRSIGS failed for
+        name: Name,
+        /// The record type in question
+        record_type: RecordType,
+    },
+
+    /// The self-signed dnskey is invalid
+    #[error("self-signed dnskey is invalid: {name}")]
+    SelfSignedKeyInvalid {
+        /// Name of the DNSKEY
+        name: Name,
+    },
+}
+
+/// The error type for dnssec errors that get returned in the crate
+#[derive(Debug, Clone, Error)]
+pub struct ProofError {
+    /// The proof derived from the failed state
+    pub proof: Proof,
+    /// The kind of error
+    pub kind: ProofErrorKind,
+}
+
+impl ProofError {
+    /// Create an error with the given Proof and Associated Error
+    pub fn new(proof: Proof, kind: ProofErrorKind) -> Self {
+        Self { proof, kind }
+    }
+
+    /// Get the kind of the error
+    pub fn kind(&self) -> &ProofErrorKind {
+        &self.kind
+    }
+}
+
+impl fmt::Display for ProofError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}: {}", self.proof, self.kind)
+    }
+}
+
+/// A wrapper type to ensure that the state of a DNSSEC proof is evaluated before use
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Proven<T> {
+    proof: Proof,
+    value: T,
+}
+
+impl<T> Proven<T> {
+    /// Wrap the value with the given proof
+    pub fn new(proof: Proof, value: T) -> Self {
+        Self { proof, value }
+    }
+
+    /// Get the associated proof
+    pub fn proof(&self) -> Proof {
+        self.proof
+    }
+
+    /// Attempts to borrow the value only if it matches flags, returning the associated proof on failure
+    ///
+    /// ```
+    /// use hickory_proto::rr::dnssec::{Proof, Proven};
+    ///
+    /// let proven = Proven::new(Proof::Bogus, 42u32);
+    ///
+    /// assert_eq!(*proven.try_borrow(Proof::Bogus).unwrap(), 42_u32);
+    /// assert_eq!(*proven.try_borrow(Proof::Bogus | Proof::Indeterminate).unwrap(), 42_u32);
+    /// assert_eq!(proven.try_borrow(Proof::Secure | Proof::Insecure).unwrap_err(), Proof::Bogus);
+    /// ```
+    pub fn try_borrow<I: Into<ProofFlags>>(&self, flags: I) -> Result<&T, Proof> {
```

Comment:
bikeshedding here really, sorry! I think the `try_borrow` and `try_take` operations make sense, but the naming isn't quite resonating with me.

`try_borrow` makes me think of `RefCell` and suggests that a failure to borrow may be some run-time borrow checking error, rather than an assertion of proof.

When working with `Option`, I tend to use `as_ref` to get at the contained reference.

I'm thinking that something like `require_proof` or perhaps simply `require` might spell things out more clearly.  Then `try_take` could be `require_proof` and `try_borrow` could be `require_proof_as_ref`?

Alternatively, `assert_proof` instead of `require_proof`?

I'm not going to die on a hill around naming; the main thing on my mind here is to avoid potential confusion with the reasonably well established `RefCell` interface.

Reply:
Yeah, I hesitated on the names, I like your ideas on better terms.

Reply:
Ok, for this one, I'm not sure what the right names are...

I propose `try_take` become `require` and `try_borrow` become `require_as_ref`. Thoughts?

Reply:
i've published an update with these names, take a look.

Reply:
@wez, any thoughts on the new changes? `#[must_use]` on `Proof`. moved to `require`, etc mentioned above?

Reply:
Sorry for the delay, have been a bit swamped recently! I've only quickly skimmed `proof.rs` here.
I think `require_xxx` is fine.
`must_use`: yeah, I think it's OK. I think it mostly helps to avoid foot-guns for users of `Proven::into_parts` as the other parts of the `Proven` interface appear to be sufficient guidance for the majority of users.